internal/xds/rbac/rbac_engine_test.go

```diff
@@ -380,226 +529,374 @@ func (s) TestRBACEngine(t *testing.T) {
 		t.Run(test.name, func(t *testing.T) {
-			// Instantiate the rbacEngine with different configurations that
+			// Instantiate the chainedRBACEngine with different configurations that are
 			// interesting to test and to query.
-			rbacEngine, err := NewEngine(test.rbacConfig)
+			cre, err := NewChainEngine(test.rbacConfigs)
 			if err != nil {
 				t.Fatalf("Error constructing RBAC Engine: %v", err)
 			}
-			// Query that created RBAC Engine with different args to see if the
-			// RBAC Engine configured a certain way works as intended.
-			for _, queryToRBACEngine := range test.rbacQueries {
-				// The matchingPolicyName returned will be empty in the case of
-				// no matching policy. Thus, matchingPolicyName can also be used
-				// to test the "error" condition of no matching policies.
-				matchingPolicyName, matchingPolicyFound := rbacEngine.FindMatchingPolicy(queryToRBACEngine.rpcData)
-				if matchingPolicyFound != queryToRBACEngine.wantMatch || matchingPolicyName != queryToRBACEngine.wantMatchingPolicyName {
-					t.Errorf("FindMatchingPolicy(%+v) returned (%v, %v), want (%v, %v)", queryToRBACEngine.rpcData, matchingPolicyName, matchingPolicyFound, queryToRBACEngine.wantMatchingPolicyName, queryToRBACEngine.wantMatch)
+			// Query the created chain of RBAC Engines with different args to see
+			// if the chain of RBAC Engines configured as such works as intended.
+			for _, data := range test.rbacQueries {
+				// Construct the context with three data points that have enough
+				// information to represent incoming RPC's. This will be how a
+				// user uses this API. A user will have to put MD, PeerInfo, and
+				// the connection the RPC is sent on in the context.
+				ctx := metadata.NewIncomingContext(context.Background(), data.rpcData.md)
+
+				// Make a TCP connection with a certain destination port. The
+				// address/port of this connection will be used to populate the
+				// destination ip/port in RPCData struct. This represents what
+				// the user of ChainEngine will have to place into
+				// context, as this is only way to get destination ip and port.
+				lis, err := net.Listen("tcp", "localhost:"+fmt.Sprint(data.rpcData.destinationPort))
+				if err != nil {
+					t.Fatalf("Error listening: %v", err)
 				}
+				connCh := make(chan net.Conn, 1)
+				go func() {
+					conn, err := lis.Accept()
+					if err != nil {
+						t.Errorf("Error accepting connection: %v", err)
+						return
+					}
+					connCh <- conn
+				}()
+				net.Dial("tcp", lis.Addr().String())
+				conn := <-connCh
+				ctx = SetConnection(ctx, conn)
+				ctx = peer.NewContext(ctx, data.rpcData.peerInfo)
+
+				err = cre.IsAuthorized(ctx, data.rpcData.fullMethod)
+				status, _ := status.FromError(err)
+				if data.wantStatusCode != status.Code() {
+					t.Fatalf("IsAuthorized(%+v, %+v) returned (%+v), want(%+v)", ctx, data.rpcData.fullMethod, status.Code(), data.wantStatusCode)
+				}
+
+				conn.Close()
+				lis.Close()
```

Comment:
`defer` close the listener right after creating it.

Reply:
I added .Close() to the fatal to avoid resources leaks. Personally (from my few months here), I find this to be cleaner than adding another .Run() call. Answer back if strongly preferred defer and I'll switch it.

Reply:
See: https://groups.google.com/g/golang-nuts/c/IvuiIvn6uXQ
I would simply introduce an anonymous function when I want to call `defer` from inside of a `for` loop.

Reply:
I like wrapping for body with anonymous function, rather than another t.Run(). Thanks for the suggestion Easwar. Did it that way.